src/compiler/crystal/repl/compiler.cr

```diff
@@ -0,0 +1,3221 @@
+      puts compiled_def.local_vars
+      puts Disassembler.disassemble(@context, compiled_def)
+      puts "=== #{target_def.owner}##{target_def.name} ==="
+    {% end %}
+
+    compiled_def
+  end
+
+  private def create_compiled_block(block : Block, target_def : Def)
+    rewrite_block_with_splat(block)
+
+    bytesize_before_block_local_vars = @local_vars.current_bytesize
+
+    @local_vars.push_block
+
+    begin
+      needs_closure_context = false
+
+      # If it's `with ... yield` we pass the "with" scope
+      # as the first block argument.
+      with_scope = block.scope
+      if with_scope
+        @local_vars.declare(WITH_SCOPE, with_scope)
+      end
+
+      block.vars.try &.each do |name, var|
+        var_type = var.type?
+        next unless var_type
+
+        if var.closure_in?(block)
+          needs_closure_context = true
+          next
+        end
+
+        next if var.context != block
+
+        @local_vars.declare(name, var_type)
+      end
+
+      if needs_closure_context
+        @local_vars.declare(Closure::VAR_NAME, @context.program.pointer_of(@context.program.void))
+      end
+
+      bytesize_after_block_local_vars = @local_vars.current_bytesize
+
+      block_args_bytesize = block.args.sum { |arg| aligned_sizeof_type(arg) }
+
+      compiled_block = CompiledBlock.new(block, @local_vars,
+        args_bytesize: block_args_bytesize,
+        locals_bytesize_start: bytesize_before_block_local_vars,
+        locals_bytesize_end: bytesize_after_block_local_vars,
+      )
+
+      # Store it so the GC doesn't collect it (it's in the instructions but it might not be aligned)
+      @context.add_gc_reference(compiled_block)
+
+      compiler = Compiler.new(@context, @local_vars,
+        instructions: compiled_block.instructions,
+        scope: @scope, def: @def, top_level: false)
+      compiler.compiled_block = @compiled_block
+      compiler.block_level = block_level + 1
+
+      compiler.compile_block(block, target_def, @closure_context)
+
+      {% if Debug::DECOMPILE %}
+        puts "=== #{target_def.owner}##{target_def.name}#block ==="
+        puts Disassembler.disassemble(@context, compiled_block.instructions, @local_vars)
+        puts "=== #{target_def.owner}##{target_def.name}#block ==="
+      {% end %}
+    ensure
+      @local_vars.pop_block
+    end
+
+    compiled_block
+  end
+
+  private def rewrite_block_with_splat(node : Block)
+    splat_index = node.splat_index
+    return unless splat_index
+
+    # If the block has a splat index, we rewrite it to something simpler.
+    #
+    # For example, assuming `y` is a tuple of 3 elements, we rewrite:
+    #
+    # ```
+    # foo do |x, *y, z|
+    #   p! x, y, z
+    # end
+    # ```
+    #
+    # to:
+    #
+    # ```
+    # foo do |x, temp1, temp2, temp3, z|
+    #   y = {temp1, temp2, temp3}
+    #   p! x, y, z
+    # end
+    # ```
+    #
+    # TODO: consider doing this in CleanupTransformer to also simplify
```

Comment:
Yes, though I'd rather not. I don't want to accidentally break compiled Crystal.

Reply:
I believe this would be easy to test, wouldn't it? There's no explicit danger for breaking things, at least not more than with any other compiler change, right?

Reply:
My idea with this branch was to modify as little as possible from compiled crystal code. Right now that's the case. If we want to start making changes across the interpreter and the compiler, I'd prefer to merge this PR first, then do those changes in separate PRs.

To be clear: I won't be making more changes to this PR until we merge it. If you or others want to continue doing changes to this PR, feel free to do so (I ran out of time, sorry)